crates/autopilot/src/run_loop.rs

```diff
@@ -429,6 +429,74 @@ impl RunLoop {
         Ok(())
     }
 
+    /// Wait for either the settlement transaction to be mined or the driver
+    /// returned a result.
+    async fn wait_for_settlement(
+        &self,
+        driver: &infra::Driver,
+        tag: &[u8],
+        request: settle::Request,
+    ) -> Result<H256, SettleError> {
+        match futures::future::select(
+            Box::pin(self.wait_for_settlement_transaction(tag, self.submission_deadline)),
+            Box::pin(driver.settle(&request, self.max_settlement_transaction_wait)),
+        )
+        .await
+        {
+            futures::future::Either::Left((res, _)) => res,
+            futures::future::Either::Right((res, _)) => {
+                res.map_err(SettleError::Failure).map(|tx| tx.tx_hash)
+            }
+        }
+    }
+
+    /// Tries to find a `settle` contract call with calldata ending in `tag`.
+    ///
+    /// Returns None if no transaction was found within the deadline or the task
+    /// is cancelled.
+    async fn wait_for_settlement_transaction(
+        &self,
+        tag: &[u8],
+        max_blocks_wait: u64,
+    ) -> Result<H256, SettleError> {
+        let start_offset = MAX_REORG_BLOCK_COUNT;
+        let current = self.eth.current_block().borrow().number;
+        let start = current.saturating_sub(start_offset);
+        let deadline = current.saturating_add(max_blocks_wait);
+        tracing::debug!(%current, %start, %deadline, ?tag, "waiting for tag");
+        let mut seen_transactions: HashSet<H256> = Default::default();
+        loop {
+            if self.eth.current_block().borrow().number > deadline {
+                break;
+            }
+            let Ok(mut hashes) = self
+                .persistence
+                .recent_settlement_tx_hashes(start..deadline + 1)
```

Comment:
I like that you are using the existing event indexing instead of trying to inspect new blocks consecutively (which makes the code simpler)

However, the settlements table also contains the `auction_id` as a column. It might require exposing a new method on persistence, but I think it would be even simpler and more coherent to just see if we can fetch a settlement by auction and return its tx_hash.

Reply:
How come I missed that column 🙈 Thanks, updated the PR.